### PR TITLE
using  LinearAllreduce flag for qwen2_moe and qwen3_moe

### PR DESCRIPTION
**Pull request by ranzhejiang**
Using LinearAllreduce flag which can adapt training, inference and eval, and get right output.

### CODE REVIEW

File: optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 1]
```diff
@@ -545,53 +545,32 @@ def gaudi_qwen2moe_block_sparse_moe_forward(self, hidden_states: torch.Tensor) -
...
-            from deepspeed import comm as dist
-
-            if dist.is_initialized():
-                dist.all_reduce(final_hidden_states, op=dist.ReduceOp.SUM)
+    experts_range = range(self.num_experts)
+    w1_list = [self.experts[i].gate_proj.weight.squeeze() for i in experts_range]
+    w2_list = [self.experts[i].down_proj.weight.squeeze() for i in experts_range]
+    w3_list = [self.experts[i].up_proj.weight.squeeze() for i in experts_range]
+
+    final_hidden_states = torch.ops.hpu.mixture_of_experts(
```
Comment (yafshar) [inline optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py:564]: @ranzhejiang is the backward path in `torch.ops.hpu.mixture_of_experts` is functional and working? The last time I tested that on `synapse 1.21.2` it was not working correctly and it was still WIP.

File: optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 2]
```diff
@@ -545,53 +545,32 @@ def gaudi_qwen2moe_block_sparse_moe_forward(self, hidden_states: torch.Tensor) -
...
+        experts_max=(self.num_experts - 1),
+    )
+    final_hidden_states = final_hidden_states.reshape(-1, sequence_length, hidden_dim)
+
+    if is_deepspeed_available() and not self.training:
+        from deepspeed import comm as dist
+        from deepspeed.module_inject.layers import LinearAllreduce
+
+        LINEAR_ALLREDUCE_TYPE = LinearAllreduce
+        if dist.is_initialized() and any(isinstance(module, LINEAR_ALLREDUCE_TYPE) for _, module in self.named_modules()):
```
Comment (yafshar) [inline optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py]:
```suggestion
        if dist.is_initialized() and any(isinstance(module, LinearAllreduce) for _, module in self.named_modules()):
```

why do you create an alias? Is there a special need for that?

File: optimum/habana/transformers/models/qwen3_moe/modeling_qwen3_moe.py [hunk 3]
```diff
@@ -636,10 +636,12 @@ def forward(self, hidden_states: torch.Tensor) -> Tuple[torch.Tensor, torch.Tens
...
         htcore.mark_step()
 
-        if not self.training and is_deepspeed_available() and self.moe_intermediate_size != w1_list[0].size(0):
+        if not self.training and is_deepspeed_available():
             from deepspeed import comm as dist
+            from deepspeed.module_inject.layers import LinearAllreduce
 
-            if dist.is_initialized():
+            LINEAR_ALLREDUCE_TYPE = LinearAllreduce
+            if dist.is_initialized() and any(isinstance(module, LINEAR_ALLREDUCE_TYPE) for _, module in self.named_modules()):
```
Comment (yafshar) [inline optimum/habana/transformers/models/qwen3_moe/modeling_qwen3_moe.py]:
```suggestion
            if dist.is_initialized() and any(isinstance(module, LinearAllreduce) for _, module in self.named_modules()):
```

same as above!

**Comment by yafshar** [comment]
@ranzhejiang can you add the tests that you have done to the PR README both for training and inference.

File: optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 4]
```diff
@@ -545,53 +545,32 @@ def gaudi_qwen2moe_block_sparse_moe_forward(self, hidden_states: torch.Tensor) -
...
+        w2=w3_list,  # Note that there is a different naming convention of w1, w2, and w3 between optimum habana's mixtral model and dynamic MoE kernel.
+        w3=w2_list,
+        permuted_weights=True,
+        activation="silu",
+        experts_min=0,
+        experts_max=(self.num_experts - 1),
+    )
+    final_hidden_states = final_hidden_states.reshape(-1, sequence_length, hidden_dim)
+
+    if is_deepspeed_available() and not self.training:
```
Comment (yafshar) [inline optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py]:
```suggestion
    if not self.training and is_deepspeed_available():
```

Re-ordered `self.training` first to short-circuit earlier

File: optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py (hunk 1 above)
Comment (Wei-Lin-Intel) [reply optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py:564]: @yafshar torch.ops.hpu.mixture_of_experts is already functional and perf. ready in 1.21.0. We have verified in both OH + DeepSpeed and Megatron-LM.

File: optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py (hunk 2 above)
Comment (ranzhejiang) [reply optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py]: Have fixed following your advice, thanks for review

File: optimum/habana/transformers/models/qwen3_moe/modeling_qwen3_moe.py (hunk 3 above)
Comment (ranzhejiang) [reply optimum/habana/transformers/models/qwen3_moe/modeling_qwen3_moe.py]: alse fixed

File: optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py (hunk 4 above)
Comment (ranzhejiang) [reply optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py]: Fixed

**Comment by ranzhejiang** [comment]
> @ranzhejiang can you add the tests that you have done to the PR README both for training and inference.

I added more tests commrnts in PR README, thanks for review

**Comment by Wei-Lin-Intel** [comment]
Please do not use PT_HPU_ENABLE_LAZY_COLLECTIVES=true for DeepSpeed training as this is an inference flag. @yafshar

File: optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 5]
```diff
@@ -545,53 +545,31 @@ def gaudi_qwen2moe_block_sparse_moe_forward(self, hidden_states: torch.Tensor) -
...
+        experts_max=(self.num_experts - 1),
+    )
+    final_hidden_states = final_hidden_states.reshape(-1, sequence_length, hidden_dim)
+
+    if not self.training and is_deepspeed_available():
+        from deepspeed import comm as dist
+        from deepspeed.module_inject.layers import LinearAllreduce
+
+        if dist.is_initialized() and any(isinstance(module, LinearAllreduce) for _, module in self.named_modules()):
+            dist.all_reduce(final_hidden_states, op=dist.ReduceOp.SUM)
```
Comment (yafshar) [inline optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py]: the above block is repeated in the forward function!

This check can become a performance bottleneck, especially for large models, since `self.named_modules()` is relatively expensive to call repeatedly during inference.

Just a suggestion: You could cache this logic using functools.lru_cache like so:

```python
    @lru_cache(None)
    def _is_deepspeed_initialized(self) -> bool:
        if not is_deepspeed_available():
            return False
        from deepspeed import comm as dist
        from deepspeed.module_inject.layers import LinearAllreduce
        return (
            dist.is_initialized() and
            any(isinstance(m, LinearAllreduce) for _, m in self.named_modules())
        )
```

Then in your forward method, you can simply call:

```py
        if not self.training and self._is_deepspeed_initialized():
            from deepspeed import comm as dist
            
            dist.all_reduce(final_hidden_states, op=dist.ReduceOp.SUM)
```

If you do not like the suggestion please find other ways, you can also break the loop as soon  there is an instance of LinearAllreduce, you do not need to go through the end

File: optimum/habana/transformers/models/qwen3_moe/modeling_qwen3_moe.py [hunk 6]
```diff
@@ -636,10 +636,11 @@ def forward(self, hidden_states: torch.Tensor) -> Tuple[torch.Tensor, torch.Tens
         )
         htcore.mark_step()
 
-        if not self.training and is_deepspeed_available() and self.moe_intermediate_size != w1_list[0].size(0):
+        if not self.training and is_deepspeed_available():
             from deepspeed import comm as dist
+            from deepspeed.module_inject.layers import LinearAllreduce
 
-            if dist.is_initialized():
+            if dist.is_initialized() and any(isinstance(module, LinearAllreduce) for _, module in self.named_modules()):
```
Comment (yafshar) [inline optimum/habana/transformers/models/qwen3_moe/modeling_qwen3_moe.py]: the same as above!

**Comment by yafshar** [comment]
@ranzhejiang thanks for the contribution. Please address the comments if possible.

Testing your PR, I noticed that there is almost 20% increase in max memory allocation which is understandable using `torch.ops.hpu.mixture_of_experts`, but more than 30%  Train Runtime improvement compared to the main branch and 60% improvement in Samples/sec. The loss difference is small less than 0.2% and I assume we can ignore it.

File: optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py (hunk 5 above)
Comment (ranzhejiang) [reply optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py]: I will check your method, thanks for your suggestion

Comment (ranzhejiang) [reply optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py]: I have changed  it following your advice

Comment (yafshar) [reply optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py]: thanks

**Comment by yafshar** [comment]
@ranzhejiang can you please sync with main and also do `make style`

**Comment by ranzhejiang** [comment]
> @ranzhejiang can you please sync with main and also do `make style`

ok, I will format it and sync with main today

**Comment by ranzhejiang** [comment]
> > @ranzhejiang can you please sync with main and also do `make style`
> 
> ok, I will format it and sync with main today

I have rebased and synced with the main branch, and also formatted some files.

**Review by yafshar — APPROVED** [review]
LGTM!

@regisss this PR is ready for your final review.

**Comment by ranzhejiang** [comment]
> The code quality check failed, please run `make style`.

done
